src/sentry/services/hybrid_cloud/organization/impl.py

```diff
@@ -512,6 +515,61 @@ def delete_option(self, *, organization_id: int, key: str) -> None:
         orm_organization = Organization.objects.get_from_cache(id=organization_id)
         orm_organization.delete_option(key)
 
+    def upsert_replicated_auth_provider(
+        self, *, auth_provider: RpcAuthProvider, region_name: str
+    ) -> None:
+        try:
+            with enforce_constraints(
+                transaction.atomic(router.db_for_write(RegionReplicatedAuthProvider))
+            ):
+                organization = Organization.objects.get(id=auth_provider.organization_id)
+                existing = RegionReplicatedAuthProvider.objects.filter(
+                    auth_provider_id=auth_provider.id
+                ).first()
+                update = {
+                    "organization": organization,
+                    "provider": auth_provider.provider,
+                    "config": auth_provider.config,
+                    "default_role": auth_provider.default_role,
+                    "default_global_access": auth_provider.default_global_access,
+                    "allow_unlinked": auth_provider.flags.allow_unlinked,
+                    "scim_enabled": auth_provider.flags.scim_enabled,
+                }
+
+                if not existing:
+                    RegionReplicatedAuthProvider.objects.create(
+                        auth_provider_id=auth_provider.id, **update
+                    )
```

Comment:
Yes!  In fact, I need to write a better test.  
In the integrity error case we want to retry on the off chance that we need the ordering of the outboxes to be a specific way to resolve organization reassignment in face of the unique constraint.
In this case (unlike the slug case), coalescing won't cause a problem because the shard identifiers will always be unique, thus the reassignment of the organization won't be ordered.  Letting integrity errors force a retry allows the ordering to be implicit -- let entropy do the work!